Add example import configuration for DSpace xoai format

alexklbuckley opened the pull request:
Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley committed:
Upstream a xsl file to convert DSpace xoai to Solr compatible format

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley commented:
We are using a custom xsl file to harvest DSpace XOAI format into VuFind Solr. 

As there is no current xsl file for XOAI we would like to upstream it so other libraries can use it.

I've removed custom elements of the xsl - I have left in the mapping of dc.rights to the edition Solr field. Please do let us know if we should remove that from this upstream file. Or if there are any other changes you would like us to make.

demiankatz commented on import/xsl/dspace-xoai.xsl:
```diff
@@ -0,0 +1,164 @@
+    detailed in the LICENSE and NOTICE files at the root of the source
+    tree and available online at
+    http://www.dspace.org/license/
+	Developed by DSpace @ Lyncode <dspace@lyncode.com>
```
Is this header comment accurate? Seems like a DSpace-specific comment that no longer applies in the context of VuFind (e.g. we don't have a NOTICE file). Does this need to be revised to indicate that it is derived from something created by Lyncode, rather than to say that it is created by Lyncode?

demiankatz replied:
Also, http://www.dspace.com/license is a dead link.

alexklbuckley replied:
thanks @demiankatz I have replaced this with a header comment more inline with the one in dspace-dim.xsl ( https://github.com/vufind-org/vufind/blob/dev/import/xsl/dspace-dim.xsl )

demiankatz commented on import/xsl/dspace-xoai.xsl:
```diff
@@ -0,0 +1,164 @@
+            <!-- INSTITUTION -->
+	    <field name="institution">
+		<xsl:if test="//*[@name='publisher']">
+	            <xsl:value-of select="//*[@name='publisher']/*/*[@name='value']"/>
```
Usually institution refers to the institution hosting DSpace -- in our other examples it's hard-coded to use a property. Is there a reason why you're putting the publisher here?

alexklbuckley replied:
thanks @demiankatz that was a mistake on my part. I've fixed it up in my followup.

demiankatz commented on import/xsl/dspace-xoai.xsl:
```diff
@@ -0,0 +1,164 @@
+            <!-- CO AUTHOR -->
+	    <xsl:for-each select="//*[@name='contributor']/*[@name='advisor']/*[@name='none']/*[@name='value']">
+               <field name="author2">
+
```
Unnecessary blank line.

alexklbuckley replied:
Thanks this is in my follow-up.

demiankatz commented on import/xsl/dspace-xoai.xsl:
```diff
@@ -0,0 +1,164 @@
+               </xsl:if>
+            </xsl:for-each>
+            <!-- CO AUTHOR -->
+	    <xsl:for-each select="//*[@name='contributor']/*[@name='advisor']/*[@name='none']/*[@name='value']">
```
Is "advisor" the only possible value here that could apply to author2?

alexklbuckley replied:
Hi @demiankatz I'm trying to find the answer to this. On the AUT DSpace instance it does look like 'advisor' is the only possible value here. I am checking for a wider DSpace context if that is the case.

demiankatz commented on import/xsl/dspace-xoai.xsl:
```diff
@@ -0,0 +1,164 @@
+		   <xsl:value-of select="//*[@name='type']"/>
+               </field>
+            </xsl:if>
+	    <!-- Rights - stored in the rights solr field for want of a better place -->
```
This says "stored in the rights" but I think you mean "edition." However, my suggestion would be to use a dynamic field for this, like `rights_str`. Then you're not putting unexpected values into an existing field, but you make the data available to custom code.

alexklbuckley replied:
Thanks this is in my follow-up.

demiankatz commented on import/xsl/dspace-xoai.xsl:
```diff
@@ -0,0 +1,164 @@
+	    <xsl:for-each select="//*[text()='ORIGINAL']/following-sibling::*[@name='bitstreams']/*[@name='bitstream']">
+               <field name="url">
+		   <xsl:value-of select="./*[@name='url']"/> # <xsl:value-of select="./*[@name='name']" />
+
```
Unnecessary blank line.

alexklbuckley replied:
Thanks this is in my follow-up.

demiankatz reviewed (changes requested):
Thanks, @alexklbuckley! See below for a few comments. I also think it would be helpful to either add the matching properties file, or else put a comment in one of the existing DSpace-related properties files raising awareness of the existence of this alternate XSLT.

It also looks to me like this XSLT probably won't work correctly with a collection of records, though maybe I'm mistaken. Are you doing all your harvesting and indexing one record at a time? If so, you might want to consider revising so that you can do batches -- the performance is tremendously improved that way.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into upstream-xoai-xsl

alexklbuckley committed:
Add example import configuration for DSpace xoai format: Implementing
review feedback

- Removing unnecessary blank lines
- Store DSpace item rights in Solr dynamic rights_str field
- Update header comment
- Store institution variable in institution field

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley commented:
> Thanks, @alexklbuckley! See below for a few comments. I also think it would be helpful to either add the matching properties file, or else put a comment in one of the existing DSpace-related properties files raising awareness of the existence of this alternate XSLT.
> 
> It also looks to me like this XSLT probably won't work correctly with a collection of records, though maybe I'm mistaken. Are you doing all your harvesting and indexing one record at a time? If so, you might want to consider revising so that you can do batches -- the performance is tremendously improved that way.

Thanks @demiankatz I have pushed a follow-up addressing all but three of your points. Which I will be following up on in the next few days.

I am checking about if 'advisor' is the only possible value for author2.

I will also add a comment to one of the existing DSpace-related properties files raising awareness of the existence of this alternative XSLT.

You're right, the harvesting is being done one record at a time. Thank you for noting that recommendation to do it in batches! Is there any documentation that you could please point me towards for making that change?

demiankatz commented:
Thanks for the progress, @alexklbuckley! Regarding your question about supporting multiple records, you might find it helpful to look at #2034 for reference. This was the pull request where we added multi-record support to all of the existing XSLT files. The key is to create a top-level template (matching "/") that can handle a collection tag or a single record, and then creating a more carefully-scoped template to transform a single record. The main key is avoiding global selectors like `//` since that can be problematic.

I'm far from an XSLT expert, so I apologize if my advice is a bit unclear (or if I'm using the wrong language). But I was able to muddle through and create #2034 by imitating the patterns of better XSLT programmers than myself, and perhaps you can do the same. If you run into trouble, I'll help if I can! :-)

alexklbuckley commented:
> Thanks for the progress, @alexklbuckley! Regarding your question about supporting multiple records, you might find it helpful to look at #2034 for reference. This was the pull request where we added multi-record support to all of the existing XSLT files. The key is to create a top-level template (matching "/") that can handle a collection tag or a single record, and then creating a more carefully-scoped template to transform a single record. The main key is avoiding global selectors like `//` since that can be problematic.
> 
> I'm far from an XSLT expert, so I apologize if my advice is a bit unclear (or if I'm using the wrong language). But I was able to muddle through and create #2034 by imitating the patterns of better XSLT programmers than myself, and perhaps you can do the same. If you run into trouble, I'll help if I can! :-)

Thanks so much @demiankatz !

demiankatz commented:
@alexklbuckley, I'm going through old pull requests and noticed that this one hasn't moved in more than a year. Is it still on your list to finish it up? Do you need any action from my side?

alexklbuckley commented:
Hi @demiankatz apologies this PR completed fell off my radar. I'll aim to have some follow-up code on this PR by the end of this week.
